### DISCUSSION TITLE
Can DAfoam be used for three-dimensional field inversion?

### DISCUSSION THREAD

**Discussion by lpz456** [mdolab/dafoam#634] [General]
Dear Professor,

I am writing to seek your expertise and guidance on a matter related to 3D field inversion.

Recently, there has been a growing interest in utilizing the DAfoam solver for 2D field inversion within the turbulence model community. This method has shown promising results and has sparked my interest in extending it to 3D field inversion. However, I have some concerns regarding the potential challenges that may arise in this extension, particularly when dealing with large number of grids.

As the number of grids increases to millions, the dimension of design variables like the correction factor for turbulent kinetic energy production term also reaches millions. Consequently, I am concerned about whether this could pose practical limitations in computing the adjoint equations due to memory constraints, or potentially lead to a deterioration in numerical robustness due to the discontinuity of the modified coefficient field after optimization? 

Given these concerns, I would greatly appreciate any advice or insights you could provide regarding the feasibility of  3D field inversion using the DAfoam solver.

Thank you for your assistance!

**Comment by friedenhe** [comment]
We never try to run it with millions of mesh cells, but if you use SNOPT/IPOPT as the optimizer, it should be able to handle millions of design variables. I am curious how it goes, and please keep us posted.

On a separate note, you may want to use our latest FIML interface where we combine field inversion and machine learning together. See a tutorial here: https://dafoam.github.io/mydoc_tutorials_field_inversion_ramp.html  The number of design variables is independent of the mesh size with this new setup. Please also refer to our latest POF paper from here https://www.researchgate.net/publication/379889817_Field_Inversion_Machine_Learning_Augmented_Turbulence_Modeling_for_Time-Accurate_Unsteady_Flow

**Reply by lpz456** [reply]
Thank you for your response.
I'm currently immersed in the literature review stage, and practical implementation is still a ways off. :-)

Although the theoretical feasibility of employing adjoint optimization for three-dimensional field inversion is recognized, ensuring numerical robustness during practical operation may pose a notable challenge.

In three-dimensional field inversion, I'm considering the potential of drawing insights from the free-form deformation (FFD) method. Adopting a similar approach may be crucial for achieving a smoother modified coefficient field to guarantee the numerical robustness. :-)

**Comment by friedenhe** [comment]
@lpz456, please read the POF paper above. Our latest FIML method is mesh-size independent.

**Reply by lpz456** [reply]
Thank you for your guidance. I will read the reference. 👍

**Reply by lpz456** [reply]
Dear Professor,

Thank you for recommending the article, which has been very helpful for further understanding the adjoint method. A related question arises:

The neural network-augmented model can make the design variables independent of the grid size. However, when applying a neural network-augmented model to optimization issue, it seems challenging to accurately solve ∂R/∂w (where R is the residual vector and w is the state variable vector).

Assuming the neural network augmented term is β(Feature1, Feature2, ...), it appears not easy to solve (∂R/∂β)·∂β/∂w in practice. Does this require approximating β function as a constant?

**Comment by friedenhe** [comment]
This is not a problem because we use a Jacobian-free adjoint approach (check this [paper](https://www.researchgate.net/publication/333518918_Effective_Adjoint_Approaches_for_Computational_Fluid_Dynamics)). We do NOT assume constant beta.
